### PR TITLE
Optimize output and flags

### PR DESCRIPTION
**Pull request by YairSlobodin1**
2nd and 3rd bullets of #191

### CODE REVIEW

File: pkg/io/csvio/sg.go [hunk 1]
```diff
@@ -20,14 +20,18 @@ func (w *Writer) WriteSG(collection *ir.SGCollection, vpc string) error {
 	if err := w.w.WriteAll(sgHeader()); err != nil {
 		return err
 	}
-	for _, sgName := range collection.SortedSGNames(vpc) {
-		vpcName := ir.VpcFromScopedResource(string(sgName))
-		sgTable, err := makeSGTable(collection.SGs[vpcName][sgName], sgName)
-		if err != nil {
-			return err
+	for _, vpcName := range collection.VpcNames() {
```
Comment (zivnevo) [inline pkg/io/csvio/sg.go:23]: I'm not sure I understand why this change is needed.
If sgName always includes vpcName, don't we get the same order before and after this change?

File: cmd/subcmds/optimizeSG.go [hunk 2]
```diff
@@ -21,24 +18,11 @@ func NewOptimizeSGCommand(args *inArgs) *cobra.Command {
 		Long:  `OptimizeSG attempts to reduce the number of security group rules in a SG without changing the semantic.`,
 		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, _ []string) error {
-			return optimization(cmd, args)
+			if err := validateFlags(args); err != nil {
+				return err
+			}
```
Comment (zivnevo) [inline cmd/subcmds/optimizeSG.go:23]: If you run `validateFlags(args)` in each and every sub-command, maybe move this code to a `PersistentPreRun` in `rootCmd`.
(you'll have to add `cobra.EnableTraverseRunHooks = true` in `NewRootCommand()`)

File: cmd/subcmds/verifyFlags.go
Comment (zivnevo) [inline cmd/subcmds/verifyFlags.go:1]: Rename to `validateFlags.go` ?

File: cmd/subcmds/root.go [hunk 3]
```diff
@@ -45,25 +45,25 @@ func NewRootCommand() *cobra.Command {
...
+	// flags set for all commands
+	rootCmd.PersistentFlags().SortFlags = false
 	_ = rootCmd.MarkPersistentFlagRequired(configFlag)
-	rootCmd.MarkFlagsMutuallyExclusive(outputFileFlag, outputDirFlag)
 
+	// sub cmds
 	rootCmd.AddCommand(NewSynthCommand(args))
 	rootCmd.AddCommand(NewOptimizeCommand(args))
 
+	// disable help command. should use --help flag instead
```
Comment (zivnevo) [inline cmd/subcmds/root.go:64]: `rootCmd.CompletionOptions.HiddenDefaultCmd = true` disables the `complete` command, not `help`

File: cmd/subcmds/verifyFlags.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package subcmds
+
+import "fmt"
+
+func validateFlags(args *inArgs) error {
+	if args.outputDir != "" && args.outputFile != "" {
+		return fmt.Errorf("only one of -d and -o can be supplied")
```
Comment (zivnevo) [inline cmd/subcmds/verifyFlags.go:12]:
```suggestion
		return fmt.Errorf("specifying both -d and -o is not allowed")
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: cmd/subcmds/root.go (hunk 3 above)
Comment (YairSlobodin1) [reply cmd/subcmds/root.go:64]: do we need this?

File: pkg/io/csvio/sg.go (hunk 1 above)
Comment (YairSlobodin1) [reply pkg/io/csvio/sg.go:23]: In optimization mode SGs names are given
==> we cannot extract the VPC names
==> we are not able to extract the SG itself (`*ir.SG`)

File: pkg/ir/acl.go [hunk 5]
```diff
@@ -15,39 +15,42 @@ import (
...
-
-type ACLWriter interface {
-	WriteACL(aclColl *ACLCollection, vpc string) error
-}
-
 func (r *ACLRule) isRedundant(rules []*ACLRule) bool {
-	for _, rule := range rules {
-		if rule.mustSupersede(r) {
+	for i := range rules {
+		if rules[i].mustSupersede(r) {
```
[the remark this answers is not in the document]
Comment (YairSlobodin1) [reply pkg/ir/acl.go:53]: fixed

**Comment by YairSlobodin1** [comment]
should also change [rootCmd description](https://github.com/np-guard/vpc-network-config-synthesis/blob/main/cmd/subcmds/root.go#L42-L43). Any idea?

File: pkg/optimize/common.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package optimize
+
+import "github.com/np-guard/vpc-network-config-synthesis/pkg/ir"
+
+type Optimizer interface {
+	// read the collection from the config object file
+	ParseCollection(filename string) error
```
[the remark this answers is not in the document]
Comment (YairSlobodin1) [reply pkg/optimize/common.go]: fixed

File: cmd/subcmds/output.go [hunk 7]
```diff
@@ -94,25 +80,30 @@ func pickWriter(args *inArgs, data *bytes.Buffer) (ir.Writer, error) {
 		return csvio.NewWriter(w), nil
 	case mdOutputFormat:
 		return mdio.NewWriter(w), nil
-	case apiOutputFormat:
-		return confio.NewWriter(w, args.configFile)
-	default:
-		return nil, fmt.Errorf("bad output format: %q", args.outputFmt)
+	case jsonOutputFormat:
+		if isSynth {
```
Comment (zivnevo) [inline cmd/subcmds/output.go:84]: don't we want a json writer for optimize? It can really help you validate optimization results with the analyzer

File: pkg/io/csvio/sg.go (hunk 1 above)
Comment (zivnevo) [reply pkg/io/csvio/sg.go:23]: I still don't get it. We do get in the config file a list of SG objects, where each SG has a reference to its parent VPC. Can't we take it from there?
BTW, we should probably allow users to specify the SG to optimize as `<vpc name>/<sg name>`, for cases where two VPCs have SGs with the same name.

File: cmd/subcmds/root.go (hunk 3 above)
Comment (zivnevo) [reply cmd/subcmds/root.go:68]: yes, otherwise you'll have a `completion` subcommand in addition to `optimize` and `synth` (which I'm not quite sure how it works)

**Comment by zivnevo** [comment]
> should also change [rootCmd description](https://github.com/np-guard/vpc-network-config-synthesis/blob/main/cmd/subcmds/root.go#L42-L43). Any ideas?

"A tool for synthesizing and optimizing VPC network configurations"

File: cmd/subcmds/output.go (hunk 7 above)
Comment (YairSlobodin1) [reply cmd/subcmds/output.go:84]: #210 
should change the output algorithm (it is different than other fmt writers)

File: cmd/subcmds/root.go (hunk 3 above)
Comment (zivnevo) [reply cmd/subcmds/root.go:68]: yes, probably a cleaner solution

File: pkg/io/csvio/sg.go (hunk 1 above)
Comment (YairSlobodin1) [reply pkg/io/csvio/sg.go:23]: fixed

**Review by zivnevo — APPROVED** [review]
